### ISSUE TITLE
SafeSocketHandle.CloseAsIs hanging in finalizer thread

### ISSUE THREAD

**Issue by jaredpar** [dotnet/runtime#40301]
This issue comes from investigating test failures on this Roslyn PR https://github.com/dotnet/roslyn/pull/46510

At the conclusion of running the unit tests for VBCSCompiler server the xUnit process will refuse to exit. The xUnit output will indicate that the tests have completed running but the process itself will not exit. Attaching the debugger to the xUnit process and there are two threads of note that are still running:

**GC Finalizer**
```
System.Private.CoreLib.dll!System.Threading.SpinWait.SpinOnceCore(int sleep1Threshold) (Unknown Source:0)
System.Net.Sockets.dll!System.Net.Sockets.SafeSocketHandle.CloseAsIs(bool abortive) (Unknown Source:0)
System.Net.Sockets.dll!System.Net.Sockets.Socket.Dispose(bool disposing) (Unknown Source:0)
System.Net.Sockets.dll!System.Net.Sockets.Socket.~Socket() (Unknown Source:0)
[Native to Managed Transition] (Unknown Source:0)
```

**.NET Sockets**
```
System.Net.Sockets.dll!System.Net.Sockets.SocketAsyncEngine.EventLoop() (Unknown Source:0)
System.Net.Sockets.dll!System.Net.Sockets.SocketAsyncEngine..ctor.AnonymousMethod__14_0(object s) (Unknown Source:0)
System.Private.CoreLib.dll!System.Threading.ThreadHelper.ThreadStart(object obj) (Unknown Source:0)
[Native to Managed Transition] (Unknown Source:0)
```

The VBCSCompiler server makes heavy use of named pipes. Looking through the `Socket` on the finalizer thread I can confirm it's a Unix Domain socket related to the named pipes the compiler is creating (the path in the end point matches the paths we create in the tests). 

Unfortunately after a day of debugging I have not been able to narrow this problem down any further:

1. The problem only repros when running the entire test assembly. I ran each test class in the assembly and none of them individually reproduce the issue. Have to run them as a group. 
1. Went through a cycle of causing the hang, identifying the test which created the socket that was hung in the finalizer, disable that test, re-run the assembly. Did this for about five different tests and it had no impact on the hang. 
1. Thought this may be related to #40289 so I tried aggressively calling `NamedPipeServerStream.Dispose` on any instance that was hung in a `WaitForConnectionAsync` call. 

None of these has had any impact though. I've also been unsuccessful in constructing a more concise repro. :frowning:

More than happy to provide any info to make tracking this down easier. 

Repro Information:
- Runtime: .NET 5 Preview 7
- OS: Ubuntu 18.04, Ubuntu 18.04 via WSL2, OSX

**Comment by stephentoub** [comment]
cc: @tmds

**Comment by jaredpar** [comment]
Did my best to narrow this down to a smaller problem. Created a [branch](https://github.com/jaredpar/roslyn/tree/repro/pipe-hang) where you can see the delta between our unit tests passing and hanging. This [commit](https://github.com/jaredpar/roslyn/commit/b1bc273a5adde37ddd972accbaa2b6f7aef7549b) shows that it's simply enabling one new test to run on Linux that causes this hang. 

To repro this do the following:

```sh
> git clone https://github.com/jaredpar/roslyn -o jarepdar
> git checkout -B repro jaredpar/repro/pipe-hang
> cd src/Compilers/Server/VBCSCompilerServerTests
> dotnet build
> dotnet msbuild /t:Test
```

Couple notes:
1. About 1/2 of  the time this will cause the tests to fail, the other half it will hang in the finalizer
1. Running the new test alone is not sufficient to demonstrate the hang, have to run the entire suite (ensure why).

A brief description of what this particular test is doing:

- Opening four `NamedPipeServerStream` instances on the same pipe name
- Creates one `NamedPipeClientStream` instance that fully connects
- Dispose three of the `NamedPipeServerStream` instances, does not dispose the connected one
- Disposes the `NamedPipeClientStream`
- That means after the test completes one of the `NamedPipeServerStream` instances remains undisposed and it **appears** this is the one that ends up stuck in the finalizer thread.

**Comment by tmds** [comment]
```
System.Private.CoreLib.dll!System.Threading.SpinWait.SpinOnceCore(int sleep1Threshold) (Unknown Source:0)
System.Net.Sockets.dll!System.Net.Sockets.SafeSocketHandle.CloseAsIs(bool abortive) (Unknown Source:0)
```

Since .NET 5, `SafeHandles` are ref'ed for on-going `Socket` operations (instead of calling `DangerousGetHandle`). If we are spinning here, there is a on-going operation against the `Socket` that keeps the handle from being released. It would be interesting to identify the operation that is on-going on some other thread. It may help put together a reproducer.

Aborting the on-going `Socket` operations is attempted by the `TryUnblockSocket`. If the `Socket` doesn't have `CLOEXEC` this method no-ops. This method is best-effort. There are tests that verify abortive behavior for UDP/TCP, but no tests for domain sockets.

https://github.com/dotnet/runtime/blob/be197adc1b2915809f59b9427a3b2270dfa1b0fc/src/libraries/System.Net.Sockets/src/System/Net/Sockets/SafeSocketHandle.Unix.cs#L196-L235

Note that async `Socket` operations always nicely abort because there are no on-going `Socket` operations that keep the `SafeHandle` ref'ed.

**Comment by antonfirsov** [comment]
> It would be interesting to identify the operation that is on-going on some other thread.

I'm currently trying to figure that out, but my gut feel is that spinning in the finalizer thread is a bad idea, and we should probably rethink the way `CloseAsIs` works regardless of the outcome of the investigation.

**Comment by jaredpar** [comment]
@tmds 

> It would be interesting to identify the operation that is on-going on some other thread. It may help put together a reproducer.

Really wish I could've found a simpler repro here that would more clearly identify the operation occurring. Spent several hours on this but was unable to do so. 

More than happy to help out with any questions on the current repro.

**Comment by antonfirsov** [comment]
Looks like the issue is not caused by outstanding blocking calls. I have a smaller repro now:
https://gist.github.com/antonfirsov/ce0cb4992e115bb4d6e8dd6862fd6780

Here is what happens in my understanding:
1. `SafePipeHandle.Unix` increases the reference counter on `SocketSafeHandle`:
https://github.com/dotnet/runtime/blob/c6da68cfd7ded9333291fe37881d7d266cfa6acb/src/libraries/System.IO.Pipes/src/Microsoft/Win32/SafeHandles/SafePipeHandle.Unix.cs#L36
2. `NamedPipeClientConnectionHost` seems to leak (= not dispose) some `NamedPipeServerStream` instances. (@jaredpar I'm wondering if this only happens in the repro branch or is this a bug in the compiler server?)
3. A `Socket` finalizer is called before the finalizers of the "owner" `SafePipeHandle` and `NamedPipeServerStream`.
4. Since the reference count is not 0, `SocketSafeHandle` is not released. `CloseAsIs` will keep spinning, blocking the finalizer thread, preventing the finalization of `SafePipeHandle` (therefore the release of `SocketSafeHandle`):
https://github.com/dotnet/runtime/blob/c6da68cfd7ded9333291fe37881d7d266cfa6acb/src/libraries/System.Net.Sockets/src/System/Net/Sockets/SafeSocketHandle.cs#L106-L114

I have an idea for a PR that would remove the spinning.

**Comment by jaredpar** [comment]
> NamedPipeClientConnectionHost seems to leak (= not dispose) some NamedPipeServerStream instances. (@jaredpar I'm wondering if this only happens in the repro branch or is this a bug in the compiler server?)

This should not be happening in the product anymore. After filing the bug I was able to find a few other places these were leaking, made a push to get rid of them, and I *believe* I got them all at this point. It's possible it's still happening in the tests though.

**Comment by tmds** [comment]
> Since the reference count is not 0, SocketSafeHandle is not released. 

:+1: Makes sense.

> I have an idea for a PR that would remove the spinning.

The spinning deals with a race condition that occurs as follows:
1. Thread A refs the SafeHandle
2. Thread B calls TryUnblockSocket
3. Thread A uses the handle for a Socket operation

If Thread B does not call TryUnblockSocket again (which it does now by spinning), the operation on thread A is not aborted.

**Comment by antonfirsov** [comment]
@tmds I may miss something important but:
do we really need to abort the operation when we are doing an abortive close from a finalizer? Isn't it better to ignore it, and enforce the abortive close?

**Comment by tmds** [comment]
> do we really need to abort the operation when we are doing an abortive close from a finalizer

When the `Socket` gets closed (by the user explicitly, or on the finalizer thread) we want all threads that are performing a blocking operation on the `Socket` to unblock (and return something like `SocketError.OperationAborted`).

**Comment by tmds** [comment]
Anton, I think we could `dup` the `SafeSocketHandle` so the `SafePipeHandle` has it's own kernel-level reference. wdyt?

**Comment by antonfirsov** [comment]
> we want all threads that are performing a blocking operation on the `Socket` to unblock (and return something like `SocketError.OperationAborted`).

@tmds what will happen to those ongoing operations if we close the underlying handle?

I think we really need to distinguish between the finalizer and a normal Dispose. It already indicates a user bug, if a finalizer is being executed. Instead of being 100% correct handling blocking operations I would only aim for preventing handle leak in those cases.

I believe the way `CloseAsIs` works now in a very unsafe way:
- We assume that `!_released` indicates an ongoing blocking operation. This is not always true.
- Currently any user code relying on `DangerousRelease` in a finalizer is risking to block the finalizer thread (a fatal bug!)

I would prefer to find a way to solve this in Sockets instead of just changing the way NamedPipes are utilizing sockets. As an alternative, we can document that user code should never `DangerousAddRef` / `DangerousRelease` on `socket.SafeHandle`, but that sounds weird to me.

**Comment by tmds** [comment]
> @tmds what will happen to those ongoing operations if we close the underlying handle?

With 3.1 on-going operations didn't ref the `SafeHandle`. With 5.0 they do, and as a consequence they prevent the `SafeHandle` from getting released (that is: actually closing). So we need something to unblock the operations.

> I think we really need to distinguish between the finalizer and a normal Dispose

In case of a finalizer, we may assume there are no on-going operations, because those would prevent the `SafeHandle` from being finalized, right? So we may skip the `TryUnblockSocket` and prevent blocking the finalizer thread.

**Comment by antonfirsov** [comment]
> With 5.0 they do, and as a consequence they prevent the `SafeHandle` from getting released (that is: actually closing)

What I mean is that in a finalizer (instead of waiting for the handle to be released) we may want force-close the handle, independently of the ref-count/release mechanism.

> In case of a finalizer, we may assume there are no on-going operations, because those would prevent the `SafeHandle` from being finalized, right?

Unfortunately, this is not true. In this particular case we are in the finalizer of the `Socket` instance, so ongoing P/Invoke-s have no effect on whether it's executed or not.

**Comment by antonfirsov** [comment]
I just noticed this comment in `TryUnblockSocket`:
> // Calling 'close' on a socket that has pending blocking calls (e.g. recv, send, accept, ...)
> // **may block indefinitely.** This is a best-effort attempt to not get blocked and make those operations return.

What does "indefinitely" mean here? Is it only for the time the blocking socket call is executing? Is it guaranteed to eventually return? Since it's a rare corner-case, in a finalizer it might be acceptable to wait for a blocking call. (As an alternative of of keeping the behavior where `SafeSocketHandle.DangerousAddRef` is risking a finalizer thread hang.)

**Comment by antonfirsov** [comment]
Looks like `SafeSocketHandle.DangerousAddRef` became unsafe with this PR:
https://github.com/dotnet/corefx/pull/38499

I wonder if the concern of indefinite spinning (or blocking `close`) was discussed in that PR /cc @wfurt

**Comment by antonfirsov** [comment]
Just an illustration of the bigger issue we have here. It's this easy to create a finalizer thread hang now:

```C#
static void Main(string[] args)
{
    CreateSocket();
    Console.WriteLine("Dying!");
    GC.Collect();
}

[MethodImpl(MethodImplOptions.NoInlining)]
private static void CreateSocket()
{
    Socket socket = new Socket(SocketType.Stream, ProtocolType.Tcp);
    bool dummy = false;
    socket.SafeHandle.DangerousAddRef(ref dummy);
}
```

**Comment by geoffkizer** [comment]
This is definitely related to the fix we made in dotnet/corefx#38499. Before that we didn't really have a socket finalizer, so we would not have hit this issue.

That said, we cannot undo the change in dotnet/corefx#38499 (at least not without making other, extensive changes). The reasoning is explained here: https://github.com/dotnet/runtime/issues/29327#issuecomment-500660152

I think the best fix for 5.0 is to simply not try to cancel outstanding operations when the Socket finalizer is called, as was suggested above. It's possible that this could lead to some unexpected behavior in weird corner cases, but I think those corner cases are restricted to the case where you do a DangerousAddRef without a corresponding DangerousRelease. Since we are currently hanging the finalizer thread in that case, we can probably live with a few quirky corner cases for now.

I do think we should revisit all of this in the future. Ideally we shouldn't need to have a finalizer on Socket. But changing that would take some effort and research etc.

**Comment by tmds** [comment]
The spinning was introduced as part of https://github.com/dotnet/corefx/pull/38804. In that PR, `Socket` PInvoke changed to use `SafeHandles` instead of `DangerousAddRef`.

Because previously there were no out-standing references, the handle was released as soon as the `Socket` got Closed. On Windows, this caused on-going operations to abort, but on Linux those operations would just keep on blocking.

Due to the use of `SafeHandle`, the handle release no longer occurs on `Close` when there are on-going operations. That is became the responsibility of `TryUnblockSocket`.

The spinning was added to deal with the race described in https://github.com/dotnet/runtime/issues/40301#issuecomment-681072351. But I did not consider someone other than `Sockets` operations could `DangerousAddRef`, which leads to infinite spinning.

Based on what was said I think we are safe to not call `TryUnblockSocket` when Closing on a finalizer thread, because there should be no on-going `Socket` operations that need to be aborted.
